File: 1. Fast and slow pointers/question 142.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        t=h=head
        cycle=False
        while h!=None and h.next!=None:
            t=t.next
            h=h.next.next
            if t==h:
                cycle=True
                break
        if not cycle:
            return None
        t2=head
        while t2!=h:
            t2=t2.next
            h=h.next
        return t2

File: 1. Fast and slow pointers/question 142.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var detectCycle = function(head) {
    hare = head;
    tortoise = head;

    while (hare !== null && hare.next !== null){
        hare = hare.next.next;
        tortoise = tortoise.next;

        if (hare === tortoise){
            temp = head;
            while (temp != hare){
                temp = temp.next;
                hare = hare.next;
            }
            return hare;
        }
    }
    return null;
};